File: typescript/src/com/murex/tbw/MainRepository.ts
import { Repository } from "./storage/Repository";
import { JsonRepository } from "./storage/JsonRepository";

export class MainRepository {
  private static runningRepository: Repository;

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  private constructor() {}

  public static configuredRepository(): Repository {
    if (!MainRepository.runningRepository) {
      MainRepository.runningRepository = new JsonRepository();
    }

    return MainRepository.runningRepository;
  }

  /* Working effectively with Legacy Code
   * https://www.goodreads.com/book/show/44919.Working_Effectively_with_Legacy_Code
   */
  // TESTING ONLY
  public static override(newRepository: Repository): void {
    MainRepository.runningRepository = newRepository;
  }

  // TESTING ONLY
  public static reset(): void {
    MainRepository.runningRepository = new JsonRepository();
  }
}

File: scala/src/main/scala/com/murex/tbw/MainRepository.scala
package com.murex.tbw

import com.murex.tbw.storage.{JsonRepository, Repository}

object MainRepository {
  private var runningRepository: Repository = _

  def configuredRepository(): Repository = {
    if (runningRepository == null) {
      runningRepository = new JsonRepository()
    }
    runningRepository
  }

  /* Working effectively with Legacy Code
   * https://www.goodreads.com/book/show/44919.Working_Effectively_with_Legacy_Code
   */
  // TESTING ONLY
  def `override`(newRepository: Repository): Unit = runningRepository = newRepository

  // TESTING ONLY
  def reset(): Unit = runningRepository = new JsonRepository()
}
